onnxruntime/core/providers/cuda/cu_inc/bitmask.cuh

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/shared_inc/fast_divmod.h"
+
+/**
+ * These functions MUST be called with an unroll factor that evenly divides the number of threads in a warp (32).
+ * In addition, this kernel MUST be launched with a number of threads in a thread block which is evenly
+ * divisible by the number of threads in a warp (32).
+ *
+ * Take unroll factor of 4 as example, we take the following approach (for threads in the first warp, that is):
+ *
+ * Thread 0 generates output booleans 0-3
+ * Thread 1 generates output booleans 4-7
+ * ...
+ * Thread 7 generates output booleans 28-31
+ *
+ * These threads all agree on the same thread mask by determining which output bitmask index they want to write to.
+ * Threads 0-7 will generate the same thread mask (for output index 0), threads 8-15 will generate the same thread mask
+ * (for output index 1), and so on.
+ *
+ * After (partially before) agreeing upon which threads will collaborate to write out a single index,
+ * each thread generates 4 random values, and shifts them into the right location in the output uint32_t.
+ * For instance:
+ *
+ * Thread 0 will perform a shift of 0
+ * Thread 1 will perform a shift of 4
+ * Thread 2 will perform a shift of 8
+ * ...
+ *
+ * For index 0, this gives us the following composition of random bits (number represents which thread generated it):
+ *
+ * 77776666555544443333222211110000
+ *
+ * After each thread shifts its bits into the right location, we broadcast the reduced value to all threads. Finally,
+ * we just choose a single thread (in our case, we choose the thread with 0 shift, but any thread from 0-7 would work
+ * for the 0-7 group).
+ *
+ * Keep in mind that this must not be conditionally called, as all threads in the warp (that haven't already exited)
+ * must reach these function calls an equal number of times, or else UB is invoked.
+ *
+ * We conditionally update the local thread's mask (with the "li < N" check), but all active threads always collaborate
+ * on the reduced value.
+ */
+
+namespace onnxruntime {
+namespace cuda {
+
+// Bitmask tensor is uint_32 type.
+using BitmaskElementType = uint32_t;
+constexpr int kNumBitsPerBitmaskElement = std::numeric_limits<BitmaskElementType>::digits;
+
+template <int NumUnroll>
+__device__ __forceinline__ void SetBitmask(const CUDA_LONG id, const CUDA_LONG mask_element_count,
+                                           const fast_divmod fdm_bits_per_element, BitmaskElementType thread_bitmask,
+                                           BitmaskElementType* mask_data) {
+  int bitmask_idx, bitmask_shift;
+  fdm_bits_per_element.divmod(id, bitmask_idx, bitmask_shift);
+  BitmaskElementType bitmask = (thread_bitmask << bitmask_shift);
+#if defined(USE_CUDA) && __CUDA_ARCH__ >= 800
```

Comment:
do we need similar condition for ROCM accordingly?

Reply:
ROCm doesn't have the new APIs (even only works on A100 for CUDA). Even the old ones are different between CUDA and ROCm. The good part is we already have WARP_SHFL_DOWN to handle the difference in ORT.